test/e2e/certificates_test.go

```diff
@@ -0,0 +1,112 @@
+// rotation is triggered, WMCO detects the new CA certificate, merge it with the initial bundle, and copy the resulting
+// bundle to all Windows nodes.
+//
+// This test accounts for the kubelet CA automatic rotation that takes place at 80% of the initial certificate's
+// lifecycle (1 year).  Note that the old certificate is still present in the cluster until day 365, so clients may be
+// able to use it.
+func (tc *testContext) testKubeletCARotation(t *testing.T) {
+	// force CA rotation
+	err := tc.forceKubeApiServerCertificateRotation()
+	require.NoError(t, err)
+	// loop all Windows nodes
+	for _, winNode := range gc.allNodes() {
+		// spin-off subtests to ensure the CA bundle file is verified in all Windows nodes
+		t.Run("node/"+winNode.Name, func(t *testing.T) {
+			err := tc.waitForKubeletCACertificateInNode(&winNode)
+			assert.NoErrorf(t, err, "kubelet CA certificate should be present in node %S", winNode.Name)
+		})
+	}
+}
+
+// forceKubeApiServerCertificateRotation forces the certificate rotation by setting the `certificate-not-after`
+// annotation to nil in the `kube-apiserver-to-kubelet-signer` secret
+func (tc *testContext) forceKubeApiServerCertificateRotation() error {
+	// note that the slash had to be encoded as ~1 because it's a reference: https://www.rfc-editor.org/rfc/rfc6901#section-3
+	patchData := fmt.Sprint(`[{"op":"replace","path":"/metadata/annotations/auth.openshift.io~1certificate-not-after","value": null }]`)
+	// invoke request
+	_, err := tc.client.K8s.CoreV1().Secrets("openshift-kube-apiserver-operator").Patch(context.TODO(),
+		"kube-apiserver-to-kubelet-signer", types.JSONPatchType, []byte(patchData), meta.PatchOptions{})
+	return err
+}
+
+// waitForKubeletCACertificateInNode waits for the kubelet CA certificate to be present in the CA bundle file for the
+// given Windows node. After retry.Interval runs an SSH job to fetch the CA bundle, if the job fails or
+// kubelet CA is not present in the node, retries every retry.Interval until retry.Timeout is reached.
+func (tc *testContext) waitForKubeletCACertificateInNode(node *core.Node) error {
+	addr, err := controllers.GetAddress(node.Status.Addresses)
+	if err != nil {
+		return err
+	}
+	// CA bundle location in Windows node. i.e. "C:\k\kubelet-ca.crt"
+	caBundlePath := windows.GetK8sDir() + nodeconfig.KubeletClientCAFilename
+	// PowerShell command to fetch content in the file
+	command := fmt.Sprintf("Get-Content -Raw -Path %s", caBundlePath)
+	// wait retry.Interval and verify the CA bundle content, try if needed
+	return wait.Poll(retry.Interval, retry.Timeout, func() (bool, error) {
+		// invoke command
+		bundleContent, err := tc.runPowerShellSSHJob("kubelet-ca-bundle-content", command, addr)
+		if err != nil {
+			// retry
+			log.Printf("error fetching CA bundle in node %s with address %s, retrying in %s. %v",
+				node.Name, addr, retry.Interval, err)
+			return false, nil
```

Comment:
Is there a reason to retry if there's a powershell job failure? The command is simple enough that it is expected to work, and if it doesn't we should return error.
```suggestion
			return false, errors.Wrapf(err, "error fetching CA bundle in node %s with address %s", node.Name, addr)
```

Reply:
A network/connectivity issue, for instance. A retry may be helpful.

Reply:
Fair enough, fine with me